### ISSUE TITLE
Gazebo-Sim Tutorial Build Error

### ISSUE THREAD

**Issue by BobbieZain** [ros-controls/gz_ros2_control#283]
**Describe the bug**
Unable to build the ROS 2 package for integrating the ros2_control controller package with docker

**To Reproduce**
1. Go to '[tutorial](https://control.ros.org/humble/doc/gz_ros2_control/doc/index.html)'
2. Run 
`cd Dockerfile`
`sudo docker build -t gz_ros2_control .`
3. See error

**Screenshots**
![image](https://github.com/ros-controls/ros2_control_demos/assets/128392753/4047f0b2-4b38-4ed1-addc-32c9f34e7fa1)

**Environment (please complete the following information):**
 - OS: Ubuntu 22.04
 - Version Humble

**Comment by christophfroehlich** [comment]
@BobbieZain Can you check #284? but how did you pass the colcon build step with the wrong branch?

**Comment by BobbieZain** [comment]
![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/7ad73582-580c-4e9c-82c6-6e1dde050941)
the problem persists

**Comment by christophfroehlich** [comment]
The linked PR has not been merged yet, please checkout the correct branch before you run the docker command.

**Comment by BobbieZain** [comment]
![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/b93cc7c2-4ce0-41aa-85e9-4f5f73f6593c)
it has built correctly, but this step doesn't seem to work
![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/12bd0f10-1510-4477-8180-399857865e2e)

**Comment by christophfroehlich** [comment]
There is another issue, #286 
`docker run -it --rm --name gz_ros2_control_demo --net host gz_ros2_control ros2 launch ign_ros2_control_demos cart_example_position.launch.py gui:=false`
works perfectly on my machine.

**Comment by BobbieZain** [comment]
Turns out the previous issue comes from not sourcing the setup.bash. The docker build runs normally now, but gives the error message about qt.qpa and rcl
![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/35b0e0c3-49e0-436a-8892-b6f8213ae2c1)
![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/3bd92fe7-b4d5-44b2-b2ee-9c091dba8a35)

**Comment by BobbieZain** [comment]
> Turns out the previous issue comes from not sourcing the setup.bash. The docker build runs normally now, but gives the error message about qt.qpa and rcl ![image](https://private-user-images.githubusercontent.com/128392753/325739736-35b0e0c3-49e0-436a-8892-b6f8213ae2c1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwODA1MjIsIm5iZiI6MTcxNDA4MDIyMiwicGF0aCI6Ii8xMjgzOTI3NTMvMzI1NzM5NzM2LTM1YjBlMGMzLTQ5ZTAtNDM2YS04ODkyLWI2ZjgyMTNhZTJjMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQyMTIzNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNzU4YWY1YTlmZTlmNzg0OGY5MDRkNDNiNWRjOTViZmRjNzU3Mjc3YjYzNjBjN2NlYjc1YzBjZDEzYWVmY2UxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bIQf_QKgwcixShrcEa1z8RCDOtgoIBMlBDJBiYlQELw) ![image](https://private-user-images.githubusercontent.com/128392753/325740056-3bd92fe7-b4d5-44b2-b2ee-9c091dba8a35.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwODA1MjIsIm5iZiI6MTcxNDA4MDIyMiwicGF0aCI6Ii8xMjgzOTI3NTMvMzI1NzQwMDU2LTNiZDkyZmU3LWI0ZDUtNDRiMi1iMmVlLTljMDkxZGJhOGEzNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQyMTIzNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NmNlYjJhNzkwZWEyNjIwMDczZjA5NjQ2YzYzMzk0MGZkNDM4MDYzNzE2MmZlMjI0MTJkMmEwZDAxOWU1ZTFmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9qVn_tJuqPSEcta3w2anabmIOWnUE0qFpMJm6CjQtes)

The simulation is still unable to run

**Comment by BobbieZain** [comment]
@christophfroehlich This code 
`sudo docker run -it --rm --name gz_ros2_control_demo --net host gz_ros2_control ros2 launch ign_ros2_control_demos cart_example_position.launch.py gui:=false`
Doesn't launch gazebo as expected.

Also could you please suggest some solutions regarding this error of 
`Could not contact service /controller_manager/load_controller` which appears when I try to  

![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/94ac1d75-8965-4a71-8737-9b7a3053e89e)

**Comment by christophfroehlich** [comment]
The "could not contact controller manager" error of the spawners just means that gz_ros2_control was not started properly. You have to read the earlier error messages in the log.

On which system are you trying to run the docker?
The qt errors mean that gui:=false was not used and it tries to launch the gui.

**Comment by BobbieZain** [comment]
`[ruby $(which ign) gazebo-1] [Err] [SystemLoader.cc:94] Failed to load system plugin [ign_ros2_control-system] : couldn't find shared library.`
I think this is what's causing the issue and I didn't manage to fix it, 
`sudo apt-get update` and `sudo apt-get upgrade` couldn't fix the problem, and I do have `ros-humble-ros2-controllers` and `ros-humble-ros2-control` installed

And this is the `docker system info`
![image](https://github.com/ros-controls/gz_ros2_control/assets/128392753/95100ba2-ae0f-4c84-864f-375cb927f7b4)

**Comment by christophfroehlich** [comment]
I just tried it again. It build but I was not able to run gazebo without gui with the given examples, see #302 

However, with the change in the gz_args to use `-s` it launches as expected for me

```
# ros2 launch ign_ros2_control_demos cart_example_position.launch.py 
[INFO] [launch]: All log files can be found below /root/.ros/log/2024-05-10-21-21-43-347429-L1CDS09-925
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [ruby $(which ign) gazebo-1]: process started with pid [926]
[INFO] [robot_state_publisher-2]: process started with pid [928]
[INFO] [create-3]: process started with pid [931]
[robot_state_publisher-2] [INFO] [1715376103.500200731] [robot_state_publisher]: got segment cart
[robot_state_publisher-2] [INFO] [1715376103.500258483] [robot_state_publisher]: got segment slideBar
[robot_state_publisher-2] [INFO] [1715376103.500263286] [robot_state_publisher]: got segment world
[create-3] [INFO] [1715376103.517451505] [ros_gz_sim]: Requesting list of world names.
[create-3] [INFO] [1715376103.837955295] [ros_gz_sim]: Requested creation of entity.
[create-3] [INFO] [1715376103.838049399] [ros_gz_sim]: OK creation of entity.
[INFO] [create-3]: process has finished cleanly [pid 931]
[INFO] [ros2-4]: process started with pid [980]
[ruby $(which ign) gazebo-1] [INFO] [1715376105.640239355] [GazeboSimROS2ControlPlugin]: robot_param_node is robot_state_publisher
[ruby $(which ign) gazebo-1] [INFO] [1715376105.640267448] [GazeboSimROS2ControlPlugin]: robot_param_node is robot_description
[ruby $(which ign) gazebo-1] [INFO] [1715376105.655809833] [gz_ros2_control]: [ign_ros2_control] Fixed joint [world_to_base] (Entity=15)] is skipped
[ruby $(which ign) gazebo-1] [INFO] [1715376105.720307810] [gz_ros2_control]: connected to service!! robot_state_publisher asking for robot_description
[ruby $(which ign) gazebo-1] [INFO] [1715376105.721735000] [gz_ros2_control]: Received URDF from param server
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726505068] [gz_ros2_control]: The position_proportional_gain has been set to: 0.1
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726608181] [gz_ros2_control]: Loading joint: slider_to_cart
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726625638] [gz_ros2_control]: 	State:
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726634638] [gz_ros2_control]: 		 position
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726690808] [gz_ros2_control]: 			 found initial value: -5.000000
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726701236] [gz_ros2_control]: 		 velocity
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726728617] [gz_ros2_control]: 		 effort
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726744757] [gz_ros2_control]: 	Command:
[ruby $(which ign) gazebo-1] [INFO] [1715376105.726751406] [gz_ros2_control]: 		 position
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727127141] [resource_manager]: Initialize hardware 'IgnitionSystem' 
[ruby $(which ign) gazebo-1] [WARN] [1715376105.727143833] [gz_ros2_control]: On init...
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727340080] [resource_manager]: Successful initialization of hardware 'IgnitionSystem'
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727393407] [resource_manager]: 'configure' hardware 'IgnitionSystem' 
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727397453] [gz_ros2_control]: System Successfully configured!
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727406352] [resource_manager]: Successful 'configure' of hardware 'IgnitionSystem'
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727409584] [resource_manager]: 'activate' hardware 'IgnitionSystem' 
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727412539] [resource_manager]: Successful 'activate' of hardware 'IgnitionSystem'
[ruby $(which ign) gazebo-1] [INFO] [1715376105.727420224] [gz_ros2_control]: Loading controller_manager
[ruby $(which ign) gazebo-1] [WARN] [1715376105.742925539] [gz_ros2_control]:  Desired controller update period (0.01 s) is slower than the gazebo simulation period (0.001 s).
[ruby $(which ign) gazebo-1] [INFO] [1715376105.896627946] [controller_manager]: Loading controller 'joint_state_broadcaster'
[ruby $(which ign) gazebo-1] [INFO] [1715376105.909421704] [controller_manager]: Configuring controller 'joint_state_broadcaster'
[ruby $(which ign) gazebo-1] [INFO] [1715376105.909512986] [joint_state_broadcaster]: 'joints' or 'interfaces' parameter is empty. All available state interfaces will be published
[ros2-4] Successfully loaded controller joint_state_broadcaster into state active
[INFO] [ros2-4]: process has finished cleanly [pid 980]
[INFO] [ros2-5]: process started with pid [1011]
[ruby $(which ign) gazebo-1] [INFO] [1715376106.757494465] [controller_manager]: Loading controller 'joint_trajectory_controller'
[ruby $(which ign) gazebo-1] [WARN] [1715376106.764812967] [joint_trajectory_controller]: [Deprecated]: "allow_nonzero_velocity_at_trajectory_end" is set to true. The default behavior will change to false.
[ruby $(which ign) gazebo-1] [INFO] [1715376106.766324534] [controller_manager]: Configuring controller 'joint_trajectory_controller'
[ruby $(which ign) gazebo-1] [INFO] [1715376106.766404167] [joint_trajectory_controller]: No specific joint names are used for command interfaces. Using 'joints' parameter.
[ruby $(which ign) gazebo-1] [INFO] [1715376106.766420605] [joint_trajectory_controller]: Command interfaces are [position] and state interfaces are [position velocity].
[ruby $(which ign) gazebo-1] [INFO] [1715376106.766431214] [joint_trajectory_controller]: Using 'splines' interpolation method.
[ruby $(which ign) gazebo-1] [INFO] [1715376106.767462306] [joint_trajectory_controller]: Controller state will be published at 50.00 Hz.
[ruby $(which ign) gazebo-1] [INFO] [1715376106.768970577] [joint_trajectory_controller]: Action status changes will be monitored at 20.00 Hz.
[ros2-5] Successfully loaded controller joint_trajectory_controller into state active
[INFO] [ros2-5]: process has finished cleanly [pid 1011]

```

I'll close this issue now because the "build error" is fixed, and the issue with the gui parameter is reported in #302 

Feel free to reopen or create a new issue if you still have problems.
